Document automated checks under `o!TR/Data Processing/Automated Checks` (#30)
osu-tournament-rating/otr-docs: docs/topics/Automated-Checks.md

    B[Apply ScoreBelowMinimum flag to RejectionReason]
    C[Does the score contain invalid mods?]
    D[Apply InvalidMods flag to RejectionReason]
    E[Does the ruleset match the tournament's ruleset?]
    F[Apply RulesetMismatch flag to RejectionReason]
    PreTerm[Is the RejectionReason null?]
    TermPositive[Change VerificationStatus to PreVerified]
    TermNegative[Change VerificationStatus to PreRejected]
   
    A -- No --> B --> C
    A -- Yes --> C
    C -- Yes --> D --> E
    C -- No --> E
    E -- Yes --> PreTerm
    E -- No --> F --> PreTerm
    PreTerm -- Yes --> TermPositive
    PreTerm -- No --> TermNegative
```

## FAQ

### How can a human manually mark all entities as `Verified`?

Most of the issues which require manual intervention are at the `Match` and `Game` levels. For example, if a `Match` has too many invalid games, it will be marked as `PreRejected` and require manual intervention. The same is true for `Game`s.

For `GameScore` entities, there are very concrete rules which can easily determine whether it should be `Rejected`, for example if the `Score` value is below the minimum (and thus very likely comes from a referee in the lobby or other anomaly).

> We also have a web interface which allows reviewers to mark an entity - and all of its children - as `Verified` or `Rejected`. Generally speaking, if at a glance everything is marked as `PreVerified`, very little effort is required to manually approve these submissions. If the opposite is true, it's likely that the submission contains invalid data.

Reviewer: I'm not the best with wording but this section reads a bit odd to me

Reviewer: > Most of the issues which require manual intervention are at the `Match` and `Game` levels. For example, if a `Match` has too many invalid games, it will be marked as `PreRejected` and require manual intervention. The same is true for `Game`s.

I think I dislike the usage of "require manual intervention" as if anything _doesn't_ require manual intervention? Items have to be manually verified or rejected in some capacity regardless of their existing verification status

Reviewer: > For `GameScore` entities, there are very concrete rules which can easily determine whether it should be `Rejected`, for example if the `Score` value is below the minimum (and thus very likely comes from a referee in the lobby or other anomaly).

This "For `GameScore` entities, there are very concrete rules which can easily determine whether it should be `Rejected`" implies that what follows will be some sort of guide for how to review `GameScore`s, but then you describe the criteria for the `GameScoreCheck`, which results in a pre rejection?

Author: Basically in this section I'm attempting to state how 'easy' it is for reviewers to not have to comb through millions of scores. I'll try to reword.

Author: Reworded, please let me know your thoughts. @myssto

# Automated Checks

> The [](DataWorkerService.md) has numerous responsibilities, one of them being a data processing step known as automated
> checks. These checks are responsible for processing various portions of data depending on the current processing step
> for a particular piece of data.

Reviewer 2: why the weird line breaks :ehW:

```suggestion
The [](DataWorkerService.md) has numerous responsibilities, one of them being a data processing step known as automated checks. These checks are responsible for processing various kinds of data and moving them through different processing steps.
```

Author: meh writerside was complaining I shouldn't have listened.

> ## Core Principals

Reviewer 2: ```suggestion
## Core Principles
```

When designing this system, we did so with the following principles in mind:

1. Human reviewers have authority over whether an entity is `Verified` or `Rejected`. As such, the system will never automatically assign these designations.
2. The automatic application of the `PreRejected` status must be as accurate as possible, based on concrete rules.
3. The process must be as transparent as possible. As such, the system tracks all changes to entities in the `audit` tables. Additionally, all entities have a `RejectionReason` enum which defines a combination of reasons why it was marked as rejected by either the system or human reviewer.
4. Do not include entities which are not `Verified` in the tournament rating algorithm.

>     * This provides an added benefit of ensuring all generated statistics are valid. Even with manually submitted data, humans make mistakes. If unverified data is introduced into the rating & statistics systems, users will notice invalid statistics and the rating ladder itself will not be completely accurate.

Reviewer 2: ```suggestion
    * Even with manually submitted data, humans make mistakes. If unverified data is introduced, users may notice invalid stats or inaccurate rating calculations, and it may be easier for bad actors to influence the algorithm.
```

## Entities

The following entities are part of this processing pipeline:

* `Tournaments`
* `Matches`
* `Games`
* `GameScores`

## Statuses

Each entity has `VerificationStatus`, `ProcessingStatus`, and `RejectionReason` fields. These fields are referenced and changed by the DataWorkerService as they move through the processing flow.

### `VerificationStatus`

Each entity shares the same `VerificationStatus` type. This type contains the following statuses:

* `None`: The entity has yet to be processed automatically.
* `PreRejected`: Based on the system's rules, this entity should be rejected.
* `PreVerified`: The system did not find anything wrong, awaiting human review.
* `Rejected`: A human marked this entity as rejected.
* `Verified`: A human marked this entity as verified.

### `ProcessingStatus`

Each entity has a unique `ProcessingStatus` type associated with it. This flag is self-explanatory: it indicates how far along an entity is in the processing pipeline.

For example, consider `TournamentProcessingStatus`:

> 1. `NeedsApproval`: The tournament is submitted but waiting approval from a verifier.

Reviewer 2: ```suggestion
1. `NeedsApproval`: The tournament is submitted but awaiting approval from a verifier.
```

2. `NeedsMatchData`: Match data needs to be fetched via the osu! API.
3. `NeedsAutomationChecks`: The tournament, and all of its children, are awaiting automation checks.

> 4. `NeedsVerification`: Awaiting human review

Reviewer 2: ```suggestion
4. `NeedsVerification`: The tournament has been checked and is awaiting human review.
```

> 5. `NeedsStatCalculation`: After human review, process statistics (must be complete before it is eligible for inclusion in the rating system).

Reviewer 2: something like this?
```suggestion
5. `NeedsStatCalculation`: Statistics, such as match scores, must be processed and stored (so that the matches are ready to be inputted into the rating algorithm).
```

Author: Will go with
> Some statistics, such as game placements, must be processed and stored as the processor requires them to function.

6. `Done`: Processing is completed. `Verified` tournaments with this status are eligible for inclusion in the rating system.

### `RejectionReason`

Each entity has a custom `RejectionReason` type with various flags which may cause it to be marked as `PreRejected`. Flags can be combined with each other to form a set of reasons. For example, a `Game` could be marked as `PreRejected` by the system due to `NoScores` and `BeatmapNotPooled`.

## Flow

Automation checks are performed in the following order.

```Mermaid
flowchart LR;
    GameScore --> Game --> Match --> Tournament
```

> > This allows the parent entities to have the context of how their children faired during the automated checks process.

Reviewer 2: ```suggestion
> This allows the parent entities to have the context of how their children fared during the automated checks process.
```

> 

### Tournament

```Mermaid
flowchart TD;

>    A[Is the count of PreVerified and/or Verified matches >= 0?]

Reviewer 2: ```suggestion
   A[Is the count of PreVerified and/or Verified matches > 0?]
```
(I guess also double check whether the 80% below is inclusive or not)

Author: Seems to be inclusive

   PreTerm[Is the RejectionReason null?]
   TermPositive[Change VerificationStatus to PreVerified]
   TermNegative[Change VerificationStatus to PreRejected]
   
   A -- No --> B --> PreTerm
   A -- Yes --> C
   C -- No --> D --> PreTerm
   C -- Yes --> PreTerm
   PreTerm -- Yes --> TermPositive
   PreTerm -- No --> TermNegative
```

### Match

```Mermaid
flowchart TD;
   A[Is the count of games > 2?]
   B[Do any games besides the first 2 have a 
      RejectionReason of BeatmapNotPooled?]
   C[Apply UnexpectedBeatmapsFound to WarningFlags]
   D[Is the EndTime property equal to 
   2007-09-17-00:00:00?]
   E[Apply NoEndTime flag to RejectionReason]
   H[Is the match name structured in a typical 
      format?]
   I[Apply UnexpectedNameFormat to WarningFlags]
   J[Does the match name start with the tournament's 
      abbreviation?]
   K[Apply NamePrefixMismatch flag to RejectionReason]
   L[Is the tournament's lobby size equal to 1?]
   M[Are the games structured in a way which supports 
      conversion to TeamVS?]
   N[Attempt to convert a full set of Head to Head games to TeamVS]
   O[Apply FailedTeamVsConversion flag to RejectionReason, repeat 

>       for all child games]

Reviewer 2: for consistency with P I suppose
```suggestion
      for all games]
```

   P[Convert all games to TeamVS, mark all games as PreVerified]
   
   F[Is the count of games equal to 0?]
   G[Apply NoGames flag to RejectionReason]
   Q[What is the count of PreVerified and/or Verified games?]
   Q1[0]
   Q2[1 or 2]

>    Q3[4 or 5]

Reviewer 2: check if 5 is included in this case or the other one
```suggestion
   Q3[3 or 4]
```

Author: Q2 should be `Q2[1 to 3]` instead.

   F -- No --> Q
   F -- Yes --> G
   Q --> Q1
   Q --> Q2
   Q --> Q3
   Q --> Q4
   Q1 --> Q_A --> PreTerm
   Q2 --> Q_B --> PreTerm
   Q3 --> Q_C --> PreTerm
   Q4 --> PreTerm
   PreTerm -- Yes --> TermPositive
   PreTerm -- No --> TermNegative
```

### Game

```Mermaid
flowchart TD;
    A[Is the beatmap null?]
    B[Is there a known mappool for the tournament?]
    C[Of all games in the tournament, is the beatmap
        used exactly once?]
    D[Apply BeatmapUsedOnce to WarningFlags]
    E[Is the beatmap in the known mappool for the tournament?]
    F[Apply BeatmapNotPooled flag to RejectionReason]
    G[Is the EndTime property equal to
        2007-09-17-00:00:00?]
    H[Apply NoEndTime flag to RejectionReason]
    I[Are invalid mods present at the game level?]
    J[Apply InvalidMods flag to RejectionReason]
    K[Does the ruleset match the tournament's ruleset?]
    L[Apply RulesetMismatch flag to RejectionReason]
    M[Is the count of scores 0?]
    N[Apply NoScores flag to RejectionReason]
    O[Is the count of PreVerified and/or Verified scores 0?]
    P[Apply NoValidScores flag to RejectionReason]
    Q[Is the count of PreVerified and/or Verified scores 

>         half that of the tournament's LobbySize?]

Reviewer 2: assuming I'm understanding this correctly
```suggestion
        twice that of the tournament's LobbySize?]
```

    S -- Yes --> U
    U -- No --> V --> PreTerm
    U -- Yes --> PreTerm
    PreTerm -- Yes --> TermPositive
    PreTerm -- No --> TermNegative
```

### GameScore

```Mermaid
flowchart TD;
    A[Is the score value > 1,000?]
    B[Apply ScoreBelowMinimum flag to RejectionReason]
    C[Does the score contain invalid mods?]
    D[Apply InvalidMods flag to RejectionReason]
    E[Does the ruleset match the tournament's ruleset?]
    F[Apply RulesetMismatch flag to RejectionReason]
    PreTerm[Is the RejectionReason null?]
    TermPositive[Change VerificationStatus to PreVerified]
    TermNegative[Change VerificationStatus to PreRejected]
   
    A -- No --> B --> C
    A -- Yes --> C
    C -- Yes --> D --> E
    C -- No --> E
    E -- Yes --> PreTerm
    E -- No --> F --> PreTerm
    PreTerm -- Yes --> TermPositive
    PreTerm -- No --> TermNegative
```

## FAQ

### How can a human manually mark all entities as `Verified`?

Most of the issues which require manual intervention are at the `Match` and `Game` levels. For example, if a `Match` has too many invalid games, it will be marked as `PreRejected` and require manual intervention. The same is true for `Game`s.

> For `GameScore` entities, there are very concrete rules which can easily determine whether it should be `Rejected`, for example if the `Score` value is below the minimum.

Reviewer 2: ```suggestion
For `GameScore` entities, there are very concrete rules which can easily determine whether it should be `Rejected`, for example if the `Score` value is below the minimum (and thus very likely comes from a referee in the lobby or other anomaly).
```
